### PR TITLE
[ROCKETMQ-233] Fix pull interval issue

### PR DESCRIPTION
**Pull request by lizhanhui**
Make pullInterval effective when pullResult is NO_NEW_MESSAGE or NO_MATCHED_MESSAGE

### CODE REVIEW

File: test/src/test/java/org/apache/rocketmq/test/client/producer/tls/TlsIntegrationTestBase.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.rocketmq.test.client.producer.tls;
+
+import org.junit.BeforeClass;
+
+public class TlsIntegrationTestBase {
```
Comment (zhouxinyu) [inline test/src/test/java/org/apache/rocketmq/test/client/producer/tls/TlsIntegrationTestBase.java]: Any relation between Tls and pull interval?

Comment (lizhanhui) [inline test/src/test/java/org/apache/rocketmq/test/client/producer/tls/TlsIntegrationTestBase.java]: No, this file should have been excluded.

File: client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java [hunk 2]
```diff
@@ -273,6 +273,17 @@ public void pullMessage(final PullRequest pullRequest) {
         final long beginTimestamp = System.currentTimeMillis();
 
         PullCallback pullCallback = new PullCallback() {
+
+            private void scheduleNextPull(PullResult pullResult) {
```
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java]: Let's have it in a comment for the method.

**Comment by shroman** [comment]
Looks good, but can you please briefly outline what the problem was -- ROCKETMQ-233 does not say much.

**Comment by lizhanhui** [comment]
@shroman This issue is to reduce fruitless polling for consumer clients. If a consumer subscribes topics which ship messages no that frequent, it may stop for a while in a smart way before performing the next round of polling, as  may relieve some pressure from brokers.

File: client/src/main/java/org/apache/rocketmq/client/consumer/PullIntervalPolicy.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.rocketmq.client.consumer;
+
+public interface PullIntervalPolicy {
+
+    void update(PullStatus status);
+
+    int getPullInterval();
```
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/consumer/PullIntervalPolicy.java]: What about making it `long`? _delay_ of executePullRequestLater(...) is long.

**Comment by shroman** [comment]
@lizhanhui Got it, thanks! How about adding comments with explanations for each policy class? I think developers will appreciate that.

**Comment by dongeforever** [comment]
LGTM  @lizhanhui please help to resolve the conflicts

File: client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vongosling) [reply client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java]: correctTagsOffsetAndSchedulePull is not a good method.

File: client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPushConsumer.java [hunk 4]
```diff
@@ -231,6 +234,7 @@ public DefaultMQPushConsumer(final String consumerGroup, RPCHook rpcHook, Alloca
         this.consumerGroup = consumerGroup;
         this.allocateMessageQueueStrategy = allocateMessageQueueStrategy;
         defaultMQPushConsumerImpl = new DefaultMQPushConsumerImpl(this, rpcHook);
+        this.pullIntervalPolicy = new BoundedExponentialPullIntervalPolicy(100, 300000);
     }
```
Comment (dongeforever) [inline client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPushConsumer.java]: 300s is too long. 
For SUB latency-sensitive applications, long polling is suggested.
So fix pull interval policy is suggested, and the default pull interval should be zero.

**Comment by lindzh** [comment]
LGTM @lizhanhui please help to resolve the conflicts

**Comment by vongosling** [comment]
@lizhanhui I will close the pr, if you happened to the same question, please let me know.

**Pull request closed without merging**